src/services/analyticsTracking.js

```diff
@@ -0,0 +1,158 @@
+        ReactGA.set({ userRole: role || GAState.userRoles[0] });
+    }
+};
+
+export const addStateInfo = obj => merge(GAState, obj);
+
+export const getUserAnalyticsState = () => GAState;
+
+export const trackUserSession = async ({ egoId, _id, acceptedTerms, roles }) => {
+    let userId = egoId;
+    if (acceptedTerms && !GAState.userId) {
+        addStateInfo({ userId, personaId: _id, userRoles: roles });
+        setUserDimensions(userId, roles[0]);
+        return true;
+    } else {
+        return false;
+    }
+};
+
+export const trackUserInteraction = async eventData => {
+    const { category, action, label } = eventData;
+    setUserDimensions();
+    ReactGA.event(eventData);
+    switch (category) {
+        case 'Modals':
+            if (action === 'Open') {
+                startAnalyticsTiming(`MODAL__${label}`);
+            } else if (action === 'Close') {
+                stopAnalyticsTiming(`MODAL__${label}`, {
+                    category,
+                    variable: 'open duration',
+                    label: label || null,
+                });
+            }
+
+            break;
+        case 'Join':
+            if (action === 'Join Completed!') {
+                stopAnalyticsTiming('join process', {
+                    category,
+                    variable: 'Join process completion time',
+                    label: label || null,
+                });
+            }
+            break;
+        default:
+            break;
+    }
+};
+
+let sanitizeName = name => name.toUpperCase().replace(/\s/g, '_');
+let getTimingEventName = name => `KF_GA_TIMING_INIT_${sanitizeName(name)}`;
+
+export const startAnalyticsTiming = eventName => {
+    timingsStorage.setItem(getTimingEventName(eventName), +new Date());
+};
+
+export const stopAnalyticsTiming = (eventName, eventData) => {
+    const event = getTimingEventName(eventName);
+    const startEventTime = timingsStorage.getItem(event);
+    let duration = null;
+
+    if (startEventTime) {
+        duration = +new Date() - startEventTime;
+        timingsStorage.setItem(`KF_GA_TIMING_${sanitizeName(eventName)}`, duration);
+    }
+
+    if (eventData && duration) {
+        trackTiming(merge({ value: duration }, eventData));
+    }
+
+    return duration;
+};
+
+export const trackTiming = async eventData => {
+    setUserDimensions();
+    ReactGA.timing(
+        merge(
+            {
+                category: null,
+                variable: null,
+                value: 0, // in milliseconds
+                label: null,
+            },
+            eventData,
+        ),
+    );
+};
+
+export const trackPageView = (page, options = {}) => {
+    setUserDimensions();
+    ReactGA.set({
+        page,
+        ...options,
+    });
+    ReactGA.pageview(page);
+};
+
+export const trackExternalLink = url => {
+    ReactGA.outboundLink({ label: url },() => {} );
+}
+
+export function withPageViewTracker(WrappedComponent, options = {}) {
```

Comment:
good intuition on the HOC, but I think there may be a simpler approach to this one.

`services/history` exports the history object that react router uses, check out documentation on `listen` here: https://github.com/ReactTraining/history.

something like 
```
  history.listen((location, action) => trackPageView(...))
```

when using ^^ you may want to whitelist paths, unless you want to track everything

Reply:
yeah was wondering if there was a better approach, feel down a few holes because the HOC caused re-rendering issues